catroid/src/main/java/org/catrobat/catroid/utils/CrashReporter.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2017 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.utils;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+import android.support.annotation.VisibleForTesting;
+import android.util.Log;
+
+import com.crashlytics.android.Crashlytics;
+import com.crashlytics.android.core.CrashlyticsCore;
+import com.google.gson.Gson;
+
+import org.catrobat.catroid.BuildConfig;
+import org.catrobat.catroid.ui.MainMenuActivity;
+import org.catrobat.catroid.ui.SettingsActivity;
+
+import io.fabric.sdk.android.Fabric;
+
+public class CrashReporter {
+
+	Context context;
+	private SharedPreferences preferences;
+	private static boolean isAnonymousReportEnabled;
+	private static boolean isCrashReportEnabled;
+	private static boolean isRecoveredFromCrash;
+	public static boolean isCrashlyticsInitialized;
+
+	public static final String RECOVERED_FROM_CRASH = "RECOVERED_FROM_CRASH";
+	public static final String EXCEPTION_FOR_REPORT = "EXCEPTION_FOR_REPORT";
+
+	private static final String TAG = CrashReporter.class.getSimpleName();
+
+	public CrashReporter(Context context) {
+		this.context = context;
+		preferences = PreferenceManager.getDefaultSharedPreferences(context);
+		isAnonymousReportEnabled = preferences.getBoolean(SettingsActivity.SETTINGS_CRASH_REPORTS, false);
+		isRecoveredFromCrash = preferences.getBoolean(RECOVERED_FROM_CRASH, false);
+		setIsCrashReportEnabled(BuildConfig.CRASHLYTICS_CRASH_REPORT_ENABLED);
+		isCrashlyticsInitialized = false;
```

Comment:
Changing `isCrashlyticsInitialized` in the constructor may lead to errors.
e.g. if you create a new instance of the CrashReporter and call initializie which sets the `isCrashlyticsInitialized` to true. The next time you create a new instance, `isCrashlyticsInitialized` will be set to false even though Crashlytics has been initialized.

All flags (the state  of the CrashReporter) are stored in static members. This hints that there should be only a single CrashReporter object. Let's try to make the CrashReporter static. Start by removing the constructor and pass in a Context to `initialize`. Depending on the result, you might also store the SharedPreferences in a static variable. Then you don't need to store `isAnonymousReportEnabled` and `isRecoveredFromCrash` in a variable since you can always get them from the SharedPreferences. Maybe create helper methods to get these flags. e.g. `isAnonymousReportEnabled() { return preferences.getBoolean(SettingsActivity.SETTINGS_CRASH_REPORTS, false)}`;

You may also need to use [guard clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) in some places. e.g. `if (!initalized) return;`.

AFAIS, the only place where you can't pass the Context as parameter, is, when you manually log an exception. In all other places you can because you always passed in the Context in the constructor already.

Give it a try and let me know how it worked out.